s3/folder.go

```diff
@@ -150,6 +155,129 @@ func (folder *Folder) PutObject(name string, content io.Reader) error {
 	return folder.uploader.upload(*folder.Bucket, folder.Path+name, content)
 }
 
+func (folder *Folder) CopyObject(srcPath string, dstPath string) error {
+	if exists, err := folder.Exists(srcPath); !exists {
+		if err == nil {
+			return errors.New("object does not exist")
+		} else {
+			return err
+		}
+	}
+	source := path.Join(*folder.Bucket, folder.Path, srcPath)
+	dst := path.Join(folder.Path, dstPath)
+	input := &s3.CopyObjectInput{CopySource: &source, Bucket: folder.Bucket, Key: &dst}
+	_, err := folder.S3API.CopyObject(input)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type s3Reader struct {
+	lastBody      io.ReadCloser
+	folder        *Folder
+	maxRetries    int32
+	objectPath    string
+	storageCursor int64
+	id            string // id useful for logs, cause its shorter than filename
+}
+
+func (reader *s3Reader) getObjectRange(from, to int64) (*s3.GetObjectOutput, error) {
+	bytesRange := fmt.Sprintf("bytes=%d-", from)
+	if to != 0 {
+		bytesRange += strconv.Itoa(int(to))
+	}
+	input := &s3.GetObjectInput{
+		Bucket: reader.folder.Bucket,
+		Key:    aws.String(reader.objectPath),
+		Range:  aws.String(bytesRange),
+	}
+	tracelog.DebugLogger.Printf("GetObject [%s] with range %s", reader.id, bytesRange)
+	return reader.folder.S3API.GetObject(input)
+}
+
+
+func (reader *s3Reader) Read(p []byte) (n int, err error) {
+	tracelog.DebugLogger.Printf("s3Reader [%s] Read to buffer [%d] bytes", reader.id, len(p))
+	reconnect := false
+	if reader.lastBody == nil { // initial connect
+		reconnect = true
+	}
+	for {
+		if reconnect {
+			reconnect = false
+			connErr := reader.reconnect()
+			if connErr != nil {
+				tracelog.DebugLogger.Printf("s3Reader [%s] reconnect failed %s", reader.id, connErr)
+				return 0, connErr
+			}
+		}
+
+		n, err = reader.lastBody.Read(p)
+		tracelog.DebugLogger.Printf("s3Reader [%s] read %d, err %s", reader.id, n, err)
+		if err != nil && err != io.EOF {
+			reconnect = true
+			continue
+		}
+		reader.storageCursor += int64(n)
+		tracelog.DebugLogger.Printf("s3Reader [%s] success read", reader.id)
+		return n, err
+	}
+}
+
+func (reader *s3Reader) reconnect() error {
+	failed := int32(0)
+	backoffSleepDuration := 0.25 * 1000 * time.Millisecond
+	sleepMultiplier := 1
+
+	for  {
+		object, err := reader.getObjectRange(reader.storageCursor, 0)
+		if err != nil {
+			failed += 1
+			tracelog.DebugLogger.Printf("s3Reader [%s] reconnect failed: %s", reader.id, err)
+			if failed >= reader.maxRetries {
+				return errors.Wrap(err, "s3Reader Too much reconnecting retries")
+			}
+			time.Sleep(time.Duration(sleepMultiplier) * backoffSleepDuration)
```

Comment:
How many retries do we allow? Do we have an upper limit of retry sleep? We may need a jitter. Maybe use build-in Retrier from S3 SDK?